CORE/world_gadgets/data_engine.rb

```diff
@@ -0,0 +1,47 @@
+
+
+class DataEngine
+
+  def initialize(logger=nil)
+    @logger = logger
+  end
+
+  def load_data_from_yml(filename)
+    @logger.debug "Loading data from file #{filename}" if @logger
+    begin
+      data = YAML.load_file(filename)
+      return data
+    rescue Psych::SyntaxError => e
+      raise "ERROR: Tally could not parse this YML file!\n"\
+		        "\tMY GUESS: You have a syntax error in your .yml file\n"\
+		        "\tYML_FILE: #{filename}\n"\
+		        "\tNOTE: Could be from inherited page YML file\n"
+      "\tORIGINAL MESSAGE: #{e.message}\n\n"
+    rescue Errno::ENOENT => e
+      raise "ERROR: Tally could not parse this YML file!\n"\
+                "\tMY FIRST GUESS: The path to this YML file is incorrect.\n"\
+                "\tMY SECOND GUESS: The yml file is missing.\n"\
+                "\tYML_FILE: #{filename}\n"\
+                "\tORIGINAL MESSAGE: #{e.message}\n\n"
+    end
+  end
+
+  def get_yml_data(data_type, filename, data_name)
+    data_name = data_name.upcase.gsub(" ","_")
```

Comment:
Nope. String operations ending with a `!` return `nil`, not the object they modify so you can't chain them together.
`data_name.upcase!.gsub!()` <- throws undefined method for nil error
`data_name.upcase.gsub!()` <- Doesn't change the original value (changes the copy returned from `upcase`

this line _could_ be changed to `data_name.upcase!` followed by `data_name.gsub!()` but I'm not sure if I like that better either